File: stack/evaluate_reverse_polish_notation.py
from typing import List


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        st = []
        for token in tokens:
            if token == "+":
                a, b = st.pop(), st.pop()
                result = a + b
                st.append(result)
            elif token == "*":
                a, b = st.pop(), st.pop()
                result = a * b
                st.append(result)
            elif token == "-":
                a, b = st.pop(), st.pop()
                result = b - a
                st.append(result)
            elif token == "/":
                a, b = st.pop(), st.pop()
                result = int(b / a)
                st.append(result)
            else:
                st.append(int(token))

        return st.pop()


s = Solution()
tokens = ["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"]
print(s.evalRPN(tokens))

File: kotlin/src/main/kotlin/stack/EvaluateReversePolishNotation.kt
package org.example.stack

import java.util.Stack

fun evalRPN(tokens: Array<String>): Int {
    val validOperators = setOf("+", "-", "*", "/")
    val st = Stack<Int>()
    for (token in tokens) {
        if (!validOperators.contains(token)) {
            st.push(token.toInt())
            continue
        }
        if (validOperators.contains(token)) {
            val second = st.pop()
            val first = st.pop()
            when (token) {
                "+" -> {
                    st.push(first + second)
                }

                "*" -> {
                    st.push(first * second)
                }

                "-" -> {
                    st.push(first - second)
                }

                "/" -> {
                    st.push(first / second)
                }
            }
        }
    }
    return st.pop()
}

fun main() {
    val input = arrayOf("4", "13", "5", "/", "+")
    val output = evalRPN(tokens = input)
    println("output: $output, expected: 6")
}